src/server/graph-generation/treeTraversal.js

```diff
@@ -0,0 +1,114 @@
+/**
+    Pathway Commons Central Data Cache
+
+    Tree Traversal
+    treeTraversal.js
+
+    Purpose  : To provide functions to search and traverse a tree represented using 
+               javascript arrays
+
+    Requires : Valid tree array
+
+    @author Harsh Mistry
+    @version 1.1 2017/10/17
+**/
+
+//Returns all values mapped to a key in a given subtree
+//Requires subtree to be valid
+//Note : - A empty array is returned if no match is found
+//       - Recurse indicates if other levels of the tree should be searched
+function searchTree(subtree, key, recurse = true) {
+    var result = []
+  
+    //Loop through all level nodes
+    for (var i = 0; i < subtree.length; i++) {
+      //Push current value if it matches
+      if (subtree[i][0] == key) {
+        result.push(subtree[i][1]);
+      }
+  
+      //Recurse on subtree if one exists
+      if (subtree[i][1] instanceof Array && recurse) {
+        result.push(searchTree(subtree[i][1], key));
+      }
```

Comment:
It would be more efficient to use a queue rather than recursion, especially if there are many levels.  Track this in a ticket for later, if the data is large.

Reply:
Parser

Reply:
If the script were to recurse through the entire file there would be a lot of levels, but the graph generation script stops after around 3-4 levels for each node. Since after around 4 levels the information tends to become redundant for tooltip purposes. So max-height of the tree is always around 4.  

Its only really an issue when parsing XML files, JSON-LD and PC2 Traversal Results which are the default methods do not require recursion. 

I'll create a ticket though for future purposes.